runtime/src/drivers.rs

```diff
@@ -537,20 +535,18 @@ impl Drivers {
     /// Counts the number of non-inactive DPE contexts and returns an error
     /// if this number is greater than or equal to the active context threshold
     /// corresponding to the privilege level of the caller.
-    ///
-    /// # Arguments
-    ///
-    /// * `pl0_pauser` - Value of PL0 PAuser
-    /// * `flags` - Flags from manifest header
-    /// * `locality` - Caller's locality
-    /// * `dpe` - DpeInstance
-    /// * `check_already_exceeded` - If true, checks that the active context threshold is already exceeded.
-    pub fn is_dpe_context_threshold_exceeded(
+    pub fn is_dpe_context_threshold_exceeded(&self) -> CaliptraResult<()> {
+        Self::is_dpe_context_threshold_exceeded_helper(
+            self.persistent_data.get().manifest1.header.pl0_pauser,
+            &self.caller_privilege_level(),
+            &self.persistent_data.get().dpe,
+        )
+    }
+
+    fn is_dpe_context_threshold_exceeded_helper(
```

Comment:
I assume this is necessary because of borrow-checker issuses? It's hard to tell from reading the code what the problem might be.

Reply:
Exactly. Here is a short snippet extracted from `initialize_dpe`

```rust
        let pdata = drivers.persistent_data.get_mut(); 
        let mut crypto = DpeCrypto::new( // <=== Holds Mutable reference to drivers until dropped.
            &mut drivers.sha384,
            &mut drivers.trng,
            &mut drivers.ecc384,
            &mut drivers.hmac384,
            &mut drivers.key_vault,
            &mut pdata.fht.rt_dice_pub_key,
            key_id_rt_cdi,
            key_id_rt_priv_key,
        );

        let (nb, nf) = Self::get_cert_validity_info(&pdata.manifest1);
        let mut env = DpeEnv::<CptraDpeTypes> {
            crypto,
           . . .
        };
        let mut dpe = DpeInstance::new_auto_init(
            &mut env,
            . . .
        )
        .map_err(|_| CaliptraError::RUNTIME_INITIALIZE_DPE_FAILED)?;
```

The env / dpe objects are used to execute `DeriveContextCmd`, so it's difficult to grab a reference to drivers in between calls.

Reply:
Oh I understand now. I thought this helper was only used in the function right above it. But it's used elsewhere as well. Ok thanks for the explanation!